mettle/main.go

```diff
@@ -61,23 +61,24 @@ var opts = struct {
-		Timeout         flags.Duration          `long:"timeout" hidden:"true" description:"Deprecated, has no effect."`
-		MinDiskSpace    flags.ByteSize          `long:"min_disk_space" default:"1G" description:"Don't accept builds unless at least this much disk space is available"`
-		MemoryThreshold float64                 `long:"memory_threshold" default:"100.0" description:"Don't accept builds unless available memory is under this percentage"`
-		VersionFile     string                  `long:"version_file" description:"File containing version tag"`
-		Costs           map[string]cli.Currency `long:"cost" description:"Per-second costs to associate with each build action."`
-		Cache           CacheOpts               `group:"Options controlling caching" namespace:"cache"`
-		Storage         StorageOpts             `group:"Options controlling communication with the CAS server" namespace:"storage"`
-		Redis           RedisOpts               `group:"Options controlling connection to Redis" namespace:"redis"`
-		Queues          struct {
+		Dir               string                  `short:"d" long:"dir" default:"." description:"Directory to run actions in"`
+		NoClean           bool                    `long:"noclean" description:"Don't clean workdirs after actions complete"`
+		Name              string                  `short:"n" long:"name" description:"Name of this worker"`
+		Browser           string                  `long:"browser" description:"Base URL for browser service (only used to construct informational user messages"`
+		Lucidity          string                  `long:"lucidity" description:"URL of Lucidity server to report to"`
+		PromGateway       string                  `long:"prom_gateway" default:"" description:"Gateway URL to push Prometheus updates to."`
+		Sandbox           string                  `long:"sandbox" description:"Location of tool to sandbox build actions with"`
+		AltSandbox        string                  `long:"alt_sandbox" description:"Location of tool to sandbox build actions with that don't explicitly request it"`
+		Timeout           flags.Duration          `long:"timeout" hidden:"true" description:"Deprecated, has no effect."`
+		MinDiskSpace      flags.ByteSize          `long:"min_disk_space" default:"1G" description:"Don't accept builds unless at least this much disk space is available"`
+		MemoryThreshold   float64                 `long:"memory_threshold" default:"100.0" description:"Don't accept builds unless available memory is under this percentage"`
+		VersionFile       string                  `long:"version_file" description:"File containing version tag"`
+		Costs             map[string]cli.Currency `long:"cost" description:"Per-second costs to associate with each build action."`
+		ImmediateShutdown bool                    `long:"immediate_shutdown" default:"false" description:"True if the worker should shut down immediately on a sigterm."`
```

Comment:
This is a nice thought but I'm not sure it necessarily has to be a flag? Is there a case where we want the immediate shutdown behaviour?

Reply:
This is being set for lofi. We've had a number of complaints about waiting 10 minutes for mettle workers to shutdown.

Reply:
There could be, from my very limited systemd knowledge I feel the current setup should just terminate the worker. We have TimeoutStopSec=600 and no ExecStop so a SIGTERM _should_ be sent right away, but on occasion it counts down from 10m before doing so.

Reply:
The 10 minute wait for tasks to complete should hopefully be fixed by the `Nack()` addition, as it'll no longer wait 10 mins for the request to be acked. This immediateShutdown option should mean that we just cancel everything straight away though - it looks to me like beforehand it would have waited for the task to finish before the worker shuts down, but perhaps there's something I've missed!

Reply:
Okay let's give it a go - I'm still suspicious that there's something weird going on with the systemd environment that's causing it but hopefully this helps!